storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java

```diff
@@ -369,16 +369,21 @@ private boolean isProducerExpired(long currentTimeMs, ProducerStateEntry produce
      * Also expire any verification state entries that are lingering as unverified.
      */
     public void removeExpiredProducers(long currentTimeMs) {
-        List<Long> keys = producers.entrySet().stream()
-                .filter(entry -> isProducerExpired(currentTimeMs, entry.getValue()))
-                .map(Map.Entry::getKey)
-                .collect(Collectors.toList());
+        List<Long> keys = new ArrayList<>();
```

Comment:
If we decided to do perf improvement, should we also avoid the copy?  I think we can just iterate over the producer id entries and remove expired ones on the spot.

Reply:
Updated. There is a significant improvement!